File: python/content_factory/api/routes/campaigns.py
"""/v1/campaigns: the Create flow — validate-and-quote (preview) before any execution (20.4).

Only deliverable types whose production branches are fully implemented in this build are
accepted; everything else is refused with a plain-language reason (no logo-grid promises).
Research in this build runs against offline fixtures; the preview says so honestly.
"""

from __future__ import annotations

from typing import Any, Literal

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from content_factory.api.deps import Principal, get_db, require_role
from content_factory.db.base import new_id
from content_factory.db.models import Role
from content_factory.deliverables.dag_compiler import compile_dag
from content_factory.config.settings import DELIVERABLE_TYPES
from content_factory.schemas import content
from content_factory.services import audit
from content_factory.services.runs import start_run

router = APIRouter(prefix="/v1/campaigns", tags=["campaigns"])
EDITOR = require_role(Role.editor)

SUPPORTED: dict[str, str] = {
    "single_image_post": "static branch (artboard → PNG) is implemented",
    "infographic": "static branch is implemented",
    "cover": "static branch is implemented",
    "carousel": "per-card render cache is implemented",
    "text_post": "text package branch is implemented",
    "short_video": "audio + video branches are implemented (mock narration in this build)",
    "long_video": "audio + video branches are implemented (mock narration in this build)",
}
UNSUPPORTED_REASON: dict[str, str] = {
    "thread": "thread splitting arrives with destination packaging (phase 9)",
    "article": "the article branch (draft, link check, SEO export) is not wired into the pipeline yet",  # noqa: E501
    "newsletter": "the email branch (MJML compile, client previews) is not wired yet",
    "email_campaign": "the email branch is not wired yet",
    "audio_clip": "standalone audio packaging is not wired yet",
    "audiogram": "audiogram compositions are not wired yet",
    "image_sequence": "the sequence engine ships; pipeline wiring arrives with the anchor workflow",
}


class DeliverableChoice(BaseModel):
    type: str
    title: str = Field(min_length=1, max_length=200)
    card_count: int | None = Field(default=None, ge=2, le=20)


class CampaignBody(BaseModel):
    topic: str = Field(min_length=1, max_length=500)
    objective: str = Field(min_length=1, max_length=1000)
    deliverables: list[DeliverableChoice] = Field(min_length=1, max_length=10)
    quality: Literal["smoke", "demo"] = "demo"


def _build_campaign(body: CampaignBody, workspace_id: str) -> content.ContentCampaign:
    export = content.DestinationBinding(
        destination=content.Destination(
            destination_id="dst_export000001", platform="export", capability_revision="2026-09-01"
        ),
        visibility="export_only",
    )
    deliverables: list[content.ContentDeliverable] = []
    for choice in body.deliverables:
        if choice.type not in SUPPORTED:
            reason = UNSUPPORTED_REASON.get(choice.type, "unknown deliverable type")
            raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, f"{choice.type}: {reason}")
        common = {"deliverable_id": new_id("dlv"), "title": choice.title, "destinations": (export,)}
        if choice.type == "single_image_post":
            deliverables.append(content.SingleImagePostSpec(layout="number_led", **common))
        elif choice.type == "infographic":
            deliverables.append(content.InfographicSpec(**common))
        elif choice.type == "cover":
            deliverables.append(content.CoverSpec(**common))
        elif choice.type == "carousel":
            deliverables.append(content.CarouselSpec(card_count=choice.card_count or 5, **common))
        elif choice.type == "text_post":
            deliverables.append(content.TextPostSpec(**common))
        elif choice.type == "short_video":
            deliverables.append(content.ShortVideoSpec(target_duration_s=(15, 30), **common))
        elif choice.type == "long_video":
            deliverables.append(content.LongVideoSpec(**common))
    brief = content.ProjectBrief(
        brief_id=new_id("brf"),
        workspace_id=workspace_id,
        input_mode=content.InputMode.brief_first,
        topic=body.topic,
        objective=body.objective,
    )
    return content.ContentCampaign(
        campaign_id=new_id("cmp"),
        workspace_id=workspace_id,
        brief=brief,
        deliverables=tuple(deliverables),
    )


@router.get("/matrix")
async def deliverable_matrix(p: Principal = Depends(EDITOR)) -> dict[str, Any]:
    """Rows = content types; cells state plainly what works and what does not (and why)."""
    rows = []
    for t in DELIVERABLE_TYPES:
        rows.append(
            {
                "type": t,
                "supported": t in SUPPORTED,
                "reason": SUPPORTED.get(t) or UNSUPPORTED_REASON.get(t, ""),
                "destinations": ["export"],
            }
        )
    return {
        "rows": rows,
        "notes": [
            "Research uses offline fixtures in this build.",
            "All output is package-only export; publishing arrives with phase 9 destinations.",
        ],
    }


@router.post("/preview")
async def preview(body: CampaignBody, p: Principal = Depends(EDITOR)) -> dict[str, Any]:
    campaign = _build_campaign(body, p.workspace_id)
    dag = compile_dag(campaign)
    return {
        "campaign": campaign.model_dump(mode="json"),
        "dag": {
            "nodes": [
                {
                    "node_id": n.node_id,
                    "stage": n.stage.value,
                    "deliverable_id": n.deliverable_id,
                    "depends_on": list(n.depends_on),
                }
                for n in dag.topological()
            ],
            "pruned": [
                {"stage": nr.stage.value, "deliverable_id": nr.deliverable_id, "reason": nr.reason}
                for nr in dag.pruned
            ],
        },
        "estimates": {"external_cost_usd": 0.0, "external_calls": 0, "local_render": True},
        "notes": [
            "Research uses offline fixtures in this build; narration uses the deterministic mock voice."  # noqa: E501
        ],
    }


@router.post("", status_code=status.HTTP_202_ACCEPTED)
async def create(
    body: CampaignBody, p: Principal = Depends(EDITOR), db: AsyncSession = Depends(get_db)
) -> dict[str, Any]:
    campaign = _build_campaign(body, p.workspace_id)
    try:
        run_id = await start_run(campaign, quality=body.quality)
    except Exception as exc:
        raise HTTPException(
            status.HTTP_503_SERVICE_UNAVAILABLE, f"workflow engine unavailable: {exc}"
        ) from exc
    await audit.record(
        db,
        "campaign.create",
        actor_account_id=p.account.id,
        workspace_id=p.workspace_id,
        target_type="campaign",
        target_id=campaign.campaign_id,
        detail={"run_id": run_id, "deliverables": [d.type for d in campaign.deliverables]},
    )
    return {"run_id": run_id, "campaign_id": campaign.campaign_id}

File: apps/web/test/msw/campaigns.ts
import { http, HttpResponse } from "msw";
import type { AuditEvent, CampaignBody, CampaignPreview, DagNode, DeliverableMatrix, OperationsHealth, PrunedStage } from "../../src/api/types";

/** Mirrors GET /v1/campaigns/matrix from the control plane (same rows and reasons). */
export const MATRIX: DeliverableMatrix = {
  rows: [
    { type: "long_video", supported: true, reason: "audio + video branches are implemented (mock narration in this build)", destinations: ["export"] },
    { type: "short_video", supported: true, reason: "audio + video branches are implemented (mock narration in this build)", destinations: ["export"] },
    { type: "single_image_post", supported: true, reason: "static branch (artboard → PNG) is implemented", destinations: ["export"] },
    { type: "carousel", supported: true, reason: "per-card render cache is implemented", destinations: ["export"] },
    { type: "infographic", supported: true, reason: "static branch is implemented", destinations: ["export"] },
    { type: "text_post", supported: true, reason: "text package branch is implemented", destinations: ["export"] },
    { type: "thread", supported: false, reason: "thread splitting arrives with destination packaging (phase 9)", destinations: ["export"] },
    { type: "article", supported: false, reason: "the article branch (draft, link check, SEO export) is not wired into the pipeline yet", destinations: ["export"] },
    { type: "newsletter", supported: false, reason: "the email branch (MJML compile, client previews) is not wired yet", destinations: ["export"] },
    { type: "email_campaign", supported: false, reason: "the email branch is not wired yet", destinations: ["export"] },
    { type: "audio_clip", supported: false, reason: "standalone audio packaging is not wired yet", destinations: ["export"] },
    { type: "audiogram", supported: false, reason: "audiogram compositions are not wired yet", destinations: ["export"] },
    { type: "cover", supported: true, reason: "static branch is implemented", destinations: ["export"] },
    { type: "image_sequence", supported: false, reason: "the sequence engine ships; pipeline wiring arrives with the anchor workflow", destinations: ["export"] },
  ],
  notes: ["Research uses offline fixtures in this build.", "All output is package-only export; publishing arrives with phase 9 destinations."],
};

/** Deterministic preview: shared chain, per-deliverable chains, honest pruned stages. */
export function makePreview(body: CampaignBody): CampaignPreview {
  const deliverables = body.deliverables.map((d, i) => ({ deliverable_id: `dlv_${i + 1}`, type: d.type, title: d.title }));
  const nodes: DagNode[] = [
    { node_id: "research", stage: "research", deliverable_id: null, depends_on: [] },
    { node_id: "plan", stage: "plan", deliverable_id: null, depends_on: ["research"] },
  ];
  const pruned: PrunedStage[] = [{ stage: "ingest", deliverable_id: null, reason: "no uploads or URLs to ingest" }];
  for (const d of deliverables) {
    nodes.push({ node_id: `script:${d.deliverable_id}`, stage: "script", deliverable_id: d.deliverable_id, depends_on: ["plan"] });
    nodes.push({ node_id: `render:${d.deliverable_id}`, stage: "render", deliverable_id: d.deliverable_id, depends_on: [`script:${d.deliverable_id}`] });
    nodes.push({
      node_id: `compile_destination_packages:${d.deliverable_id}`,
      stage: "compile_destination_packages",
      deliverable_id: d.deliverable_id,
      depends_on: [`render:${d.deliverable_id}`],
    });
    if (d.type !== "short_video" && d.type !== "long_video") {
      pruned.push({ stage: "synthesize_narration", deliverable_id: d.deliverable_id, reason: "static deliverables have no narration" });
    }
  }
  return {
    campaign: { campaign_id: "cmp_test0001", deliverables },
    dag: { nodes, pruned },
    estimates: { external_cost_usd: 0, external_calls: 0, local_render: true },
    notes: ["Research uses offline fixtures in this build; narration uses the deterministic mock voice."],
  };
}

export const HEALTH: OperationsHealth = {
  ok: false,
  checks: [
    { name: "database", status: "ok", detail: "postgres reachable, migrations applied", fix: null },
    { name: "temporal", status: "ok", detail: "workflow engine reachable at 127.0.0.1:7233", fix: null },
    { name: "renderer", status: "warn", detail: "remotion bundle is older than the schema package", fix: "run just render-smoke to rebuild" },
    { name: "web push", status: "fail", detail: "VAPID keys are not configured", fix: "run ./setup.sh to generate keys" },
    { name: "gpu", status: "skip", detail: "no GPU in this environment", fix: null },
  ],
};

export const AUDIT: AuditEvent[] = [
  { id: "aud_1", action: "campaign.create", actor: "acct_1", workspace_id: "ws_1", target: "campaign:cmp_1", created_at: "2026-08-31T09:00:00Z" },
  { id: "aud_2", action: "run.approve", actor: "acct_1", workspace_id: "ws_1", target: "run:run_2", created_at: "2026-08-31T08:10:00Z" },
  { id: "aud_3", action: "session.login", actor: "acct_1", workspace_id: null, target: null, created_at: "2026-08-31T08:00:00Z" },
];

/** Defaults for the product endpoints; individual tests override what they need. */
export const productHandlers = [
  http.get("*/v1/campaigns/matrix", () => HttpResponse.json(MATRIX)),
  http.post("*/v1/campaigns/preview", async ({ request }) => HttpResponse.json(makePreview((await request.json()) as CampaignBody))),
  http.post("*/v1/campaigns", () => HttpResponse.json({ run_id: "run_new1", campaign_id: "cmp_new1" }, { status: 202 })),
  http.get("*/v1/operations/health", () => HttpResponse.json(HEALTH)),
  http.get("*/v1/operations/audit", () => HttpResponse.json(AUDIT)),
  // Web push is unconfigured by default, exactly like a fresh local build.
  http.get("*/v1/notifications/vapid-public-key", () => HttpResponse.json({ detail: "web push is not configured (run setup)" }, { status: 503 })),
  http.post("*/v1/notifications/subscriptions", () => HttpResponse.json({ id: "push_1" }, { status: 201 })),
  http.delete("*/v1/notifications/subscriptions", () => new HttpResponse(null, { status: 204 })),
  http.post("*/v1/notifications/test", () => HttpResponse.json({ sent: 1, total: 1 })),
];
